File: ssh/Game.rb
#encoding:utf-8

require_relative 'FuzzyPlayer' #PREGUNTAR

module Irrgarten
	
	class Game
		private
			@@MAX_ROUNDS=10
			@@N_ROWS=5
			@@N_COLS=5
			@@E_ROW=4
			@@E_COL=4
			
			def next_player()
				@current_player_index = (@current_player_index + 1)%@players.length()
			end
			
			def actual_direction(d)
				cr = @players[@current_player_index].get_row()
				cc = @players[@current_player_index].get_col()
				
				prefdirections = @labyrinth.valid_moves(cr,cc)
				output = @players[@current_player_index].move(d,prefdirections)
				
				return output
			end
			
			def combat(m)
				r = 0
				w = Irrgarten::GameCharacter::PLAYER
				
				pattack = @players[@current_player_index].attack();
				lose = m.defend(pattack)
				
				while (lose == false) and (r < @@MAX_ROUNDS)
					w = Irrgarten::GameCharacter::MONSTER
					r += 1
					
					mattack = m.attack()
					lose = @players[@current_player_index].defend(mattack)
					
					if lose == false
						pattack = @players[@current_player_index].attack();
						w = Irrgarten::GameCharacter::PLAYER
						lose = m.defend(pattack)
					end
				end
				
				log_rounds(r,@@MAX_ROUNDS)
				
				return w
			end
			
			def manage_reward(w)
				if (w == Irrgarten::GameCharacter::PLAYER)
					@players[@current_player_index].receive_reward()
					log_player_won()
				else
					log_monster_won()
				end
					
			end
			
			def manage_resurrection()
				resurrect = Irrgarten::Dice.resurrect_player()
				
				if resurrect == true
					@players[@current_player_index].resurrect()
					fp = FuzzyPlayer.new(@players[@current_player_index])
					@players[@current_player_index]=fp
					@labyrinth.put_fuzzy_player(fp)
					log_resurrected()
				else
					log_player_skip_turn()
				end
			end
			
			def log_player_won()
				@log = @log +"Player won \n"
			end
			
			def log_monster_won()
				@log = @log +"Monster won \n"
			end
			
			def log_resurrected()
				@log = @log +"Player resurrected \n"
			end
			
			def log_player_skip_turn()
				@log = @log +"Player lost turn \n"
			end
			
			def log_player_no_orders()
				@log = @log +"Player couldn't follow the orders \n"
			end
			
			def log_no_monsters()
				@log = @log +"Player moved to a empty square or couldn't move \n"
			end
			
			def log_rounds(r,m)
				@log = @log+r.to_s+"/"+m.to_s+ "combat rounds \n"
			end
			
		public
			def initialize(np) 
				@players = []
				@monsters = []
				@log = ""
				
				i = 0
				while i < np
					@players[i] = Irrgarten::Player.new(i.to_s,Irrgarten::Dice.random_intelligence,Irrgarten::Dice.random_strength)
					i+=1
				end
				
				@current_player_index = Irrgarten::Dice.who_starts(np)
				@labyrinth = Irrgarten::Labyrinth.new(@@N_ROWS,@@N_COLS,@@E_ROW,@@E_COL)
				
				@labyrinth.spread_players(@players)
				configure_labyrinth()
			end
			
			def finished()
				return @labyrinth.have_a_winner()
			end
			
			def next_step(d)
				@log = ""
				dead = @players[@current_player_index].dead()
				if dead == false
					direct = actual_direction(d)
					if direct != d
						log_player_no_orders()
					end
					
					monst = @labyrinth.put_player(direct,@players[@current_player_index])
					
					if monst == nil
						log_no_monsters()
					else
						winner = combat(monst)
						manage_reward(winner)
					end
				else
					manage_resurrection()
				end
				
				endgame = finished()
				
				if endgame == false
					next_player()
				end
				
				return endgame
			end
			
			def get_gamestate()
				p = ""
				i = 0
				while i < @players.length
					p = p + @players[i].to_s + " "
					i+=1
				end
				
				m = ""
				i = 0
				while i < @monsters.length
					m = m + @monsters[i].to_s + " "
					i+=1
				end
				
				gamestate = Irrgarten::GameState.new(@labyrinth.to_s,p,m,@current_player_index,@labyrinth.have_a_winner,@log)
				return gamestate
			end
			
			def configure_labyrinth()
				ns = @@N_ROWS * @@N_COLS
				
				i = 0 
				while i < ns / 5
					r = Dice.random_pos(@@N_ROWS)
					c = Dice.random_pos(@@N_COLS)
					mname ="monster"+i.to_s
					@monsters[i] = Monster.new(mname,Irrgarten::Dice.random_intelligence,Irrgarten::Dice.random_strength)
					@labyrinth.add_monster(r,c,@monsters[i])
					i+=1
				end
				
				i = 0 
				while i < (ns / 20)
					r = Dice.random_pos(@@N_ROWS)
					c = Dice.random_pos(@@N_COLS)
					@labyrinth.add_block(Irrgarten::Orientation::HORIZONTAL,r,c,@@N_ROWS)
						
					r = Dice.random_pos(@@N_ROWS)
					c = Dice.random_pos(@@N_COLS)
					@labyrinth.add_block(Irrgarten::Orientation::VERTICAL,r,c,@@N_COLS)
					i+=1
				end
			end
	end
end
	

File: irrgarten/src/irrgarten/Game.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;

/**
 * @author Juan Molina Molina
 * @author Paula Jiménez Ruiz
 */
public class Game {
    
    private static int MAX_ROUNDS=10;  // Atributo de clase
    private static int N_ROWS=5;
    private static int N_COLS=5;
    private static int E_ROW= 4;
    private static int E_COL= 4;
    public int currentPlayerIndex;
    private String log;
    
    private ArrayList<Player> players;
    private ArrayList<Monster> monsters;
    private Labyrinth labyrinth;
    
    public Game(int nplayers) {
    
        players = new ArrayList<Player>();
        monsters = new ArrayList<Monster>();
        Dice dice=new Dice();
        for (int i=0; i<nplayers; i++){
            char number = (char) (i+48);
            Player player = new Player(number, dice.randomIntelligence(), dice.randomStrength());
            players.add(player);
        }
        currentPlayerIndex = dice.whoStarts(nplayers);
        labyrinth = new Labyrinth (N_ROWS,N_COLS,E_ROW,E_COL);
        
        labyrinth.spreadPlayers(players);
        configureLabyrinth();
    }
    
    public boolean finished() {
    
        return labyrinth.haveAWinner();
    }
    
    public boolean nextStep(Directions preferredDirection) {
    
        log = "";
        boolean dead = players.get(currentPlayerIndex).dead();
        if (!dead){
            Directions direction = actualDirection(preferredDirection);
            
            if (direction != preferredDirection){
                logPlayerNoOrders();
            }
            
            Monster monster = labyrinth.putPlayer(direction, players.get(currentPlayerIndex));
            
            if (monster == null) {
                logNoMonster();
            }
            else {
                GameCharacter winner = combat(monster);
                manageReward(winner);
            }
        }
        else {
            manageResurrection();
        }
        
        boolean endGame = finished();
        
        if (!endGame){
            nextPlayer();
        }
        return endGame;
    }
    
    public GameState getGameState() {
    
        GameState gamestate = new GameState(labyrinth.toString(),players.toString(),
            monsters.toString(),currentPlayerIndex,labyrinth.haveAWinner(),log);
        return gamestate;
    }
    
    private void configureLabyrinth() {
    
        Dice dice = new Dice();
        int n_squares = N_ROWS*N_COLS;
        for (int i = 0; i<n_squares/20; i++){
            int row = dice.randomPos(N_ROWS);
            int col = dice.randomPos(N_COLS);
            labyrinth.addBlock(Orientation.HORIZONTAL, row, col,N_ROWS);
            row = dice.randomPos(N_ROWS);
            col = dice.randomPos(N_COLS);
            labyrinth.addBlock(Orientation.VERTICAL, row, col,N_COLS);
        }
        
        for (int i = 0; i<n_squares/5; i++){
            int row = dice.randomPos(N_ROWS);
            int col = dice.randomPos(N_COLS);;
            Monster monster = new Monster("Monster #"+i,dice.randomIntelligence(),dice.randomStrength());
            monsters.add(monster);
            labyrinth.addMonster(row, col, monster);
        }
        
    }
    
    public void nextPlayer() {
        
        currentPlayerIndex = (currentPlayerIndex+1)%players.size();
    }
    
    
    private Directions actualDirection(Directions preferedDirection) {
    
        int currentRow = players.get(currentPlayerIndex).getRow();
        int currentCol = players.get(currentPlayerIndex).getCol();
        
        ArrayList <Directions> preferredDirections = labyrinth.validMoves(currentRow, currentCol);
        
        Directions output =players.get(currentPlayerIndex).move(preferedDirection, preferredDirections);
        
        return output;
    }
    
    private GameCharacter combat(Monster monster) {
    
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        
        float playerAttack = players.get(currentPlayerIndex).attack();
        boolean lose = monster.defend(playerAttack);
        
        while ((!lose) && (rounds < MAX_ROUNDS)){
            winner = GameCharacter.MONSTER;
            rounds++;
            
            float monsterAttack = monster.attack();
            lose = players.get(currentPlayerIndex).defend(monsterAttack);
            
            if (!lose){
                playerAttack = players.get(currentPlayerIndex).attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        logRounds(rounds,MAX_ROUNDS);
        
        return winner;
    }
    
    private void manageReward(GameCharacter winner) {
        
        if (winner == GameCharacter.PLAYER){
            
            players.get(currentPlayerIndex).receiveReward();
            logPlayerWon();
        }
        else {
            logMonsterWon();
        }
    }
    
    private void manageResurrection() {

        Dice dice= new Dice();
        boolean resurrect = dice.resurrectPlayer();

        if(resurrect){
            players.get(currentPlayerIndex).resurrect();
            FuzzyPlayer fuzzyplayer =new FuzzyPlayer(players.get(currentPlayerIndex));
            players.set(currentPlayerIndex, fuzzyplayer);
            labyrinth.putFuzzyPlayer(fuzzyplayer);
            logResurrected();
        }
        else {
            logPlayerSkipTurn();
        }
    }
    
    private void logPlayerWon() {
    
        log = log.concat("Player won \n");
    }
    
    private void logMonsterWon() {
    
        log = log.concat("Monster won \n");
    }
    
    private void logResurrected() {
    
        log = log.concat("Player resurrected \n");
    }
    
    private void logPlayerSkipTurn() {
    
        log =log.concat("Player lost turn \n");
    }
    
    private void logPlayerNoOrders() {
    
        log = log.concat("Player couldn't follow the orders \n");
    }
    
    private void logNoMonster() {
    
        log = log.concat("Player moved to a empty square or couldn't move \n");
    }
    
    private void logRounds(int rounds, int max) {
    
        String logline = rounds+"/"+max+ " combat rounds \n";
        log = log.concat(logline);
    }
    
}
